source/xlsxreader.d

```diff
@@ -111,46 +121,46 @@ struct Sheet {
 		mixin(format(t, T, T[0].toUpper ~ T[1 .. $]));
 	}
 
-	ColumnUntyped iterateColumnUntyped(size_t col, size_t startRow, size_t endRow) @safe {
+	ColumnUntyped iterateColumnUntyped(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(T) iterateColumn(T)(size_t col, size_t startRow, size_t endRow) {
+	Column!(T) iterateColumn(T)(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(long) iterateColumnLong(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(long) iterateColumnLong(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(double) iterateColumnDouble(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(double) iterateColumnDouble(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(string) iterateColumnString(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(string) iterateColumnString(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(DateTime) iterateColumnDateTime(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(DateTime) iterateColumnDateTime(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(Date) iterateColumnDate(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(Date) iterateColumnDate(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
-	Column!(TimeOfDay) iterateColumnTimeOfDay(size_t col, size_t startRow, size_t endRow) @safe {
+	Column!(TimeOfDay) iterateColumnTimeOfDay(size_t col, size_t startRow, size_t endRow) @safe return {
 		return typeof(return)(&this, col, startRow, endRow);
 	}
 
 	// Row
 
-	Iterator!T getRow(T)(size_t row, size_t startColumn, size_t endColumn) @safe {
-		return typeof(return)(this.iterateRow!T(row, startColumn, endColumn).array); // TODO: why .array?
+	T[] getRow(T)(size_t row, size_t startColumn, size_t endColumn) @safe return {
+		return this.iterateRow!T(row, startColumn, endColumn).array;
 	}
```

Comment:
Thanks Ate for the review & nice comments, much appreciated!

---

I thought the comment referred to 'why does the Iterator constructor require an array already?'. That was my thought, before seeing that `Iterator` is just a (useless?) wrapper around an array. By getting rid of `Iterator` and returning the array directly, I think it's clearer now, and what to do when wanting to iterate lazily or avoid the copy - using the `iterate*` methods for the underlying range.

[Btw, the copy is inefficient, as the ranges don't expose a length, although they'd have one, so the `Appender` needs to grow dynamically...]

Reply:
Oh, rereading the code I think I now see why the copy was necessary. In `Sheet`, the row is an array of `Cell`s, but we're returning an array of T here, meaning we need to copy unless we return a lazy range instead.

I suspect the `Iterator` was a placeholder for such a lazy range. Underneath, it was an array, but it could have been improved to a lazy range without breaking client code, unlike a vanilla array. We're trading away future-proofing for less complexity by removing it but maybe it's justified.

We can still initialize Appender (or just the result array directly) to correct length by `endColumn - startColumn`

Reply:
> Oh, rereading the code I think I now see why the copy was necessary. In Sheet, the row is an array of Cells, but we're returning an array of T here, meaning we need to copy unless we return a lazy range instead.

No, the *typed* `iterate*` methods (`iterateRow!T()`, `iterateColumn!T()`) all return these lazy ranges, where the `Cell` from the underlying untyped range is converted to the target type. I'm not sure why we have these typed `Column/Row` range templates - to me, it looks like a straight-forward `map` case. E.g.:
```D
auto iterateColumn(T)(size_t col, size_t startRow, size_t endRow) @safe return {
    // currently: return Column!T(&this, col, startRow, endRow);
    return iterateColumnUntyped(col, startRow, endRow).map(cell => convertTo!T(cell.xmlValue));
}
```

AFAICT, the Column/Row range templates currently only seem to additionally cache the front value.

Reply:
> No, the typed iterate* methods (iterateRow!T(), iterateColumn!T()) all return these lazy ranges,

Yes, but I meant the functions (getRow, getColumn) that used to return `Iterator` and now return a slice.

(sorry for needlessly delaying a day with answering - I hope you weren't waiting for my opinion)

Reply:
I know, just saying that the lazy, non-copying ranges are already there. So I can't think of a good reason for an `Iterator!T` to be (planned as) something else than, well, an eagerly copied array.